pubnub/event_engine/effects.py

```diff
@@ -0,0 +1,431 @@
+import asyncio
+import logging
+import math
+
+from typing import Optional, Union
+from pubnub.endpoints.presence.heartbeat import Heartbeat
+from pubnub.endpoints.presence.leave import Leave
+from pubnub.endpoints.pubsub.subscribe import Subscribe
+from pubnub.enums import PNReconnectionPolicy
+from pubnub.exceptions import PubNubException
+from pubnub.features import feature_enabled
+from pubnub.models.server.subscribe import SubscribeMessage
+from pubnub.pubnub import PubNub
+from pubnub.event_engine.models import events, invocations
+from pubnub.models.consumer.common import PNStatus
+from pubnub.workers import BaseMessageWorker
+
+
+class Effect:
+    pubnub: PubNub = None
+    event_engine = None
+    invocation: Union[invocations.PNManageableInvocation, invocations.PNCancelInvocation]
+    stop_event = None
+    logger: logging.Logger
+    task: asyncio.Task
+
+    def set_pn(self, pubnub: PubNub):
+        self.pubnub = pubnub
+
+    def __init__(self, pubnub_instance, event_engine_instance,
+                 invocation: Union[invocations.PNManageableInvocation, invocations.PNCancelInvocation]) -> None:
+        self.invocation = invocation
+        self.event_engine = event_engine_instance
+        self.pubnub = pubnub_instance
+
+        self.logger = logging.getLogger("pubnub")
+
+    def run(self):
+        pass
+
+    def run_async(self, coro):
+        loop: asyncio.AbstractEventLoop = self.pubnub.event_loop
+        if loop.is_running():
+            self.task = loop.create_task(coro)
+        else:
+            self.task = loop.run_until_complete(coro)
+
+    def stop(self):
+        if self.stop_event:
+            self.logger.debug(f'stop_event({id(self.stop_event)}).set() called on {self.__class__.__name__}')
+            self.stop_event.set()
+        if hasattr(self, 'task') and isinstance(self.task, asyncio.Task) and not self.task.cancelled():
+            self.task.cancel()
+
+    def get_new_stop_event(self):
+        event = asyncio.Event()
+        self.logger.debug(f'creating new stop_event({id(event)}) for {self.__class__.__name__}')
+        return event
+
+    def calculate_reconnection_delay(self, attempts):
```

Comment:
Let's leave it for now. But this method could be the part of `ReconnectionPolicy` class, not Effect. You can improve it while working on automatic retry

Reply:
CLEN-1799